framework/tlvf/src/include/tlvf/TlvList.h

```diff
@@ -0,0 +1,136 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2016-2019 Intel Corporation
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#ifndef _TlvList_H_
+#define _TlvList_H_
+
+#include "ieee_1905_1/eTlvType.h"
+#include <memory>
+#include <vector>
+#include <tlvf/BaseClass.h>
+
+namespace ieee1905_1 {
+
+class TlvList {
+
+public:
+    TlvList();
+    ~TlvList();
+
+public:
+    template <class T> std::shared_ptr<T> addClass()
+    {
+        std::shared_ptr<T> ptr;
+        if (m_buff) {
+            if (m_class_vector.size() == 0) {
+                ptr = std::make_shared<T>(m_buff, m_parse, m_swap);
+            } else {
+                ptr = std::make_shared<T>(m_class_vector.back(), m_parse, m_swap);
+            }
+            if (!ptr || ptr->isInitialized() == false) {
+                return std::shared_ptr<T>();
+            } else {
+                m_class_vector.push_back(std::shared_ptr<BaseClass>(ptr));
+            }
+        } else {
+            return std::shared_ptr<T>();
+        }
+        return ptr;
+    }
+
+    template <class T> std::shared_ptr<T> dynamicCast(std::shared_ptr<BaseClass> ptr) const
+    {
+        return std::dynamic_pointer_cast<T>(ptr);
+    }
+
+    /**
+     * @brief Get the Class object at index idx in the all classes array
+     *
+     * @param idx index in the all classes array
+     * @return std::shared_ptr<BaseClass> to the class object at index idx, nullptr if not found
+     */
+    std::shared_ptr<BaseClass> getClass(size_t idx) const;
+
+    /**
+     * @brief Get the (first) Class object
+     *
+     * @tparam T class template
+     * @return std::shared_ptr<T> to the first object found of type T, nullptr if not found
+     */
+    template <class T> std::shared_ptr<T> getClass() const
+    {
+        for (size_t idx = 0; idx < getClassCount(); idx++) {
+            if (auto c = std::dynamic_pointer_cast<T>(getClass(idx)))
+                return c;
+        }
+        return nullptr;
+    }
+
+    /**
+     * @brief Get a class object of type T in index `idx` in the logical array containing
+     *        all classes of type T.
+     *
+     * @tparam T class template
+     * @param idx index in the class T array
+     * @return std::shared_ptr<T> to the T class at index `idx` in the class T array
+     */
+    template <class T> std::shared_ptr<T> getClass(size_t idx) const
+    {
+        size_t idx_ = 0;
+        for (size_t i = 0; i < getClassCount(); i++) {
+            if (auto c = std::dynamic_pointer_cast<T>(getClass(i))) {
+                if (idx_++ == idx)
```

Comment:
Please do the increment as a separate statement.

A simpler way could be:

```
for (auto it : m_class_vector) {
    if (auto c = std::dynamic_pointer_cast<T>(it)) {
        if (idx == 0)
            return c;
        idx--;
    }
}
```

However I'm not particularly fond of modifying arguments or decrementing an iterator myself, so probably not such a good idea :-)

Reply:
I'm not sure I agree that `idx-- == idx` is not readable enough. People should know the order of operations which happen here, if they don't then its a red sign.